### ISSUE TITLE
Sketcher: chains of geometry generate duplicate points

### ISSUE THREAD

**Issue by pierreporte** [FreeCAD/FreeCAD#13391]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Because vertex of geometric entities (lines, circles, arcs, etc.) are part of them, drawing a shape generate duplicate points at every contact between elements. For example, a rectangle would have eight points, two for each side. Mathematically, it doesn’t make sense. A rectangle has only four points and the sides connect the points. The points are not part of the sides.

In FreeCAD, the consequence is that at every intersection, there are two coincident points, thus a lot of constraints, which need to be shown to the user. It leads to sub-obtimal choices (in my opinion) like coloring coincident points (see #13098), which lead to new problems (see my comments in #13098).

An other likely consequence is that it is impossible to switch the construction status of the vertex, because they inherit the status of their parent. It may be the cause of issue #11920.

Making points independent to other geometric entities will allow to dramatically reduce the number of constraints, thus enabling to use a separate icon to display the remaining ones, like the point on object constraint.

It may also help to better indicate the constrain status of the whole sketch. Currently, a line (the segment and the two vertex) turn green only if everything is fully-constrained. It would maybe allow to color only one point or just the segment when applicable.

### Full version info

```shell
OS: Ubuntu 23.10 (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.36807 (Git) AppImage
Build type: Release
Branch: main
Hash: cc96f2718e139d670e08c4226bdea9955a102ed6
Python 3.11.8, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: French/France (fr_FR)
Installed mods: 
  * BillOfMaterials-WB 0.0.7
  * CosmeticThread3D
  * Curves 0.6.29
  * CurvedShapes 1.0.8
  * sheetmetal 0.4.6
  * OpenDark 2023.12.17
  * OpenLight 2023.12.17
```

### Subproject(s) affected?

Sketcher

### Anything else?

Inspired by discussions in #13098 and #11919.

May be linked to #11920

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by pierreporte** [comment]
@maxwxyz @PaddleStroke @FlachyJoe

**Comment by maxwxyz** [comment]
We need to make a discussion about these sketcher based UI issues somewhere for a big picture @obelisk79

**Comment by PaddleStroke** [comment]
This would be a very deep change that would probably break compatibility with older files.
I think this is a difficult project.

**Comment by pierreporte** [comment]
> This would be a very deep change that would probably break compatibility with older files. I think this is a difficult project.

It means that the best time to do it is right now. Breaking changes like these will be hard to accept after 1.0. Bigger changes would paradoxically be easier to explain.

**Comment by FlachyJoe** [comment]
Hi, if we could hide coincidence = red points can totally disappear, then we could hide the auto-created ones.

**Comment by pierreporte** [comment]
@FlachyJoe Yes but it would solve only one aspect of the problem.

**Comment by pierreporte** [comment]
@FlachyJoe This explains why it is a deep change. Can you give some examples of such constraints?

**Comment by FlachyJoe** [comment]
line length is the major

If you have to constraint something else than a rectangle parallel to axis you will need to add points.
![image](https://github.com/FreeCAD/FreeCAD/assets/10371513/625c4157-3b36-448f-bc94-388a7e0a515b)

The coincident count is not a subject: internally they have to be present and we can remove them from the list as they are now shown in the geometry (and the checkbox do nothing).

**Comment by pierreporte** [comment]
@PaddleStroke I have an idea to solve this issue. Let me know what you think about it.

There are two core aspects:

1. The vertex of geometric entities should be turned into internal geometry, but in a way that they couldn’t be restored by `Sketcher_RestoreInternalAlignmentGeometry`.
2. When the user creates geometry, when the vertex are not placed on an existing point, then a standalone point (`Sketcher_CreatePoint`) is created. For example, when a circle is created from the origin of the sketch, the circle center which would be a vertex will not be turned into a point; when a line is created in the 2D space but completely unconnected, two points will be created where its vertex are; when a second line is created with its start at the end of that line, the start vertex is constrained to the existing point, and a second one is created at the free end.

Because standalone points would be created, they could be turned freely into construction or normal geometry, effectively solving #11920.

It would be necessary to mark points created implicitly as such, so that the deletion of geometry deletes them automatically when they are only connected to the deleted geometry.

Coincident constrained generated by this mechanism should be hidden to the user, and only shown when created manually, for example when two unconnected lines become connected by their end point. This way, the coincidence constrain icon would be visible only in relevant cases (like In Catia). In this situation, the two points should be merged, and a feature should be made available to split back the point in two unconnected points at the end of each segment.

I don’t know if this is still a very big change as discussed earlier, but hopefully it’s simpler.

**Comment by pierreporte** [comment]
@PaddleStroke What’s your opinion about my above proposal?

**Comment by PaddleStroke** [comment]
I still think that this is an extremely difficult project. And I fail to understand what we would gain.
I understand that conceptually a rectangle has 4 points. But what does it matter if it has 8? Ok in some cases it happens that you delete a segment where 3 points were coïncident, and that removes a coïncidence you wanted to keep. But it's quite e a edge case.
And again I think it is close to impossible. Well not impossible of course. But that's like rewriting a lot, and I mean a lot, of sketcher's workbench. Like more than half of the workbench. I may be wrong of course.

**Comment by kaiwas** [comment]
> I may be wrong of course.

You are right. It is probably very complicated and time-consuming. But it will allow to expand the selection methods.
For a long time I could not understand why I could not assign constraints to the vertices selected by the frame. And this would be very convenient. Constant aiming with the mouse and click click click is very annoying.

![Image](https://github.com/user-attachments/assets/10491b51-44af-48b9-bafe-240aa41c50a9)

**Comment by PaddleStroke** [comment]
Ooh I see I missed that. Indeed the box selection is selecting 4 points when you use box selection.
Well we can fix that specific thing without rewriting all of sketcher. The constraints tools could be a bit smarter and detect when selection is twice more many points, then remove overlapping points in this case.

**Comment by pierreporte** [comment]
There is also the problem that vertex can't be turned into normal geometry. The proposal was made with that in mind actually.

**Comment by PaddleStroke** [comment]
First @kaiwas please create an issue about the specific issue you reported (the box selection thing).

Secondly, I think this specific ticket here should be closed as not planed (@maxwxyz). I think it relies on a misconception of how the workbench works internally. It's one of those things that appear logic from a user perspective ('rectangle should have 4 points'), but which doesn't make sense and is hard to explain why. But I'll try : 

The workbench do not create actual points for edges. The geometry vector only store the edges. Then in coin3d we are creating a visual representation for the edge end/mid-points. So that the user can select them and create constraints on them. But ultimately the constraint that is created is on the edge geometry. We just specify that it's on 'start' of edge so that the solver can understand it.

It is not possible to make a line start and end 'not exist'. They exist there is no way around it : A rectangle is really 4 lines, each line has a start and a end point, that's 8 points.
You can pretend they don't exist by not showing them visually. But what would be the point? Coincident points are already stacked and so one of them is already invisible.

**Comment by kaiwas** [comment]
> First [@kaiwas](https://github.com/kaiwas) please create an issue about the specific issue you reported (the box selection thing).

This isn't a generation issue. I don't care how many are actually generated (under the hood). The most important thing is that choosing and using them should be easy. So, everything discussed here concerns ease of selection, not generation.

I'll create a separate issue later. For now, I'm a bit tired of FreeCAD.

UPD/ This issue is extremely annoying. I'd better fix it as soon as possible.

**Comment by pierreporte** [comment]
Excerpt of https://github.com/FreeCAD/FreeCAD/issues/11920#issuecomment-3502592593 about why this may still be relevant:

> If two vertex are coincident (say a line ends at a circle center), then there would be two vertex that could be turned into construction or defining state. How do you make sure that users don’t change the wrong one? It is why I thought that current vertex should be internal so that the duplicates are hidden and standalone points are created on the fly as needed so that there is only one point to modify in this situation.

@PaddleStroke 

>  You can pretend they don't exist by not showing them visually. But what would be the point? Coincident points are already stacked and so one of them is already invisible.

The idea is to make them invisible and replaced with dynamically created standalone points.
